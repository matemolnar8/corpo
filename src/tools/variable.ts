import { tool } from "ai";
import { stringifySmall } from "../log.ts";
import z from "zod";

const variables = new Map<string, string>();

export const getVariable = (name: string) => {
  return variables.get(name);
};

export const setVariable = (name: string, value: string) => {
  variables.set(name, value);
};

export const resetVariables = () => {
  variables.clear();
};

export const variableInputSchema = z.object({
  name: z.string().describe("The name of the variable to store"),
  value: z.string().describe("The value of the variable to store"),
  overwrite: z
    .boolean()
    .default(false)
    .describe("Whether to overwrite the variable if it already exists"),
});

export const variableOutputSchema = z.object({
  success: z.boolean().describe("Whether the variable was stored successfully"),
  reason: z
    .string()
    .optional()
    .describe("The reason for the success or failure"),
});

export const storeVariableTool = tool<
  { name: string; value: string; overwrite: boolean },
  { success: boolean; reason?: string }
>({
  description: "Store a variable for future use in a workflow",
  inputSchema: variableInputSchema,
  outputSchema: variableOutputSchema,
  execute: ({ name, value, overwrite }) => {
    console.log(
      `[Custom] Running tool 'store_variable' with args: ${
        stringifySmall({ name, valueLength: value.length, overwrite })
      }`,
    );
    if (variables.has(name) && !overwrite) {
      console.log(
        `Variable '${name}' already exists and will not be overwritten`,
      );
      const output = { success: false, reason: "Variable already exists" } as const;
      console.log(`[Custom] Tool 'store_variable' completed with result: ${stringifySmall(output)}`);
      return output;
    }

    setVariable(name, value);
    console.log(`Variable '${name}' stored with value '${value.substring(0, 50)}${value.length > 50 ? "..." : ""}'`);
    const output = { success: true } as const;
    console.log(`[Custom] Tool 'store_variable' completed with result: ${stringifySmall(output)}`);
    return output;
  },
});

export const retrieveVariableTool = tool<
  { name: string },
  { success: boolean; value?: string; reason?: string }
>({
  description: "Retrieve a variable from the workflow",
  inputSchema: z.object({
    name: z.string().describe("The name of the variable to retrieve"),
  }),
  outputSchema: z.object({
    success: z
      .boolean()
      .describe("Whether the variable was retrieved successfully"),
    value: z
      .string()
      .optional()
      .describe("The value of the variable, if found"),
    reason: z
      .string()
      .optional()
      .describe("The reason for the success or failure"),
  }),
  execute: ({ name }) => {
    console.log(`[Custom] Running tool 'retrieve_variable' with args: ${stringifySmall({ name })}`);
    if (!variables.has(name)) {
      console.log(`Variable '${name}' not found`);
      const output = { success: false, reason: "Variable not found" } as const;
      console.log(`[Custom] Tool 'retrieve_variable' completed with result: ${stringifySmall(output)}`);
      return output;
    }

    const value = getVariable(name) as string;

    console.log(
      `Variable '${name}' retrieved with value '${value.substring(0, 50)}${value.length > 50 ? "..." : ""}'`,
    );
    const output = { success: true, value } as const;
    console.log(`[Custom] Tool 'retrieve_variable' completed with result: ${stringifySmall(output)}`);
    return output;
  },
});
