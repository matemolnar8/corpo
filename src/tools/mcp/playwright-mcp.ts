import { Tool, tool } from "ai";
import { z } from "zod";
import { MCPClient, MCPTool } from "./mcp-client.ts";
import { jsonSchema } from "@ai-sdk/provider-utils";
import { logger, stringifySmall } from "../../log.ts";
import type { ImageContent, TextContent } from "@modelcontextprotocol/sdk/types.js";
import { setVariable } from "../variable.ts";
import { replaceSecretsInArgsWithTracking, replaceSecretsInResultAllowed } from "../secret.ts";
import { deferPromise } from "../../utils.ts";
import { resolvePlaywrightToolDescription, withIncludeSnapshotDescription } from "./playwright-utils.ts";
import { PLAYWRIGHT_ALLOWED_TOOL_NAMES } from "./playwright-tools.ts";

const PLAYWRIGHT_MCP = {
  command: "npx",
  args: ["@playwright/mcp@latest"] as const,
} as const;

export type PlaywrightToolOutput = { content: (TextContent | ImageContent)[]; isError?: boolean };

export class PlaywrightMCP {
  private client?: MCPClient;

  constructor() {}

  async connect({ headless }: { headless?: boolean } = {}): Promise<void> {
    const client = new MCPClient({
      command: PLAYWRIGHT_MCP.command,
      args: headless ? [...PLAYWRIGHT_MCP.args, "--headless"] : [...PLAYWRIGHT_MCP.args],
    });
    await client.connect();
    logger.debug("MCP", "Connected to Playwright MCP");
    this.client = client;
  }

  async disconnect(): Promise<void> {
    if (this.client) await this.client.disconnect();
    logger.debug("MCP", "Disconnected from Playwright MCP");
  }

  private getClient(): MCPClient {
    if (!this.client) throw new Error("MCP client not connected");
    return this.client;
  }

  private filterAllowedTools(tools: MCPTool[]): MCPTool[] {
    const allowed = new Set(PLAYWRIGHT_ALLOWED_TOOL_NAMES);

    // assert that every tool name in allowed is in tools
    for (const name of allowed) {
      if (!tools.some((t) => t.name === name)) {
        throw new Error(`Tool '${name}' is not provided by Playwright MCP`);
      }
    }

    const filtered = tools.filter((t) => allowed.has(t.name));
    return filtered;
  }

  async listAllTools(): Promise<MCPTool[]> {
    const client = this.getClient();
    const { tools } = await client.listTools();
    return tools;
  }

  async listFilteredTools(): Promise<MCPTool[]> {
    const all = await this.listAllTools();
    return this.filterAllowedTools(all);
  }

  // Yay single threading!
  private previousToolPromise?: Promise<void>;

  async callTool(
    name: string,
    args: Record<string, unknown>,
    options: { includeSnapshot?: boolean } = {},
  ) {
    const originalPreviousToolPromise = this.previousToolPromise;
    const previousToolDeferred = deferPromise<void>();
    this.previousToolPromise = previousToolDeferred.promise;
    // Wait for previous tool to complete
    if (originalPreviousToolPromise) {
      logger.debug("MCP", "Waiting for previous tool to complete");
      await originalPreviousToolPromise;
    } else {
      logger.debug("MCP", "No previous tool to wait for");
    }

    // Workaround for race condition in Playwright MCP
    await new Promise((resolve) => setTimeout(resolve, 1000));

    let result;
    if (name === "browser_snapshot_and_save") {
      result = await this.snapshotAndSaveTool.execute!(args as { variable: string }, {
        messages: [],
        toolCallId: crypto.randomUUID(),
      });
    } else {
      result = await this.callMcpTool(name, args, options);
    }

    previousToolDeferred.resolve();

    return result;
  }

  async callMcpTool(
    name: string,
    args: Record<string, unknown>,
    options: { includeSnapshot?: boolean } = {},
  ) {
    const client = this.getClient();
    // Log placeholders but not resolved secret values
    logger.debug("MCP", `Tool '${name}' args: ${stringifySmall(args)}`);
    try {
      const { value: safeArgs, usedSecretNames } = replaceSecretsInArgsWithTracking(args);
      const result = await (client.callTool(name, safeArgs) as Promise<PlaywrightToolOutput>);
      const filtered = options.includeSnapshot ? result : this.removeSnapshots(result);
      const masked = replaceSecretsInResultAllowed(filtered, usedSecretNames);
      logger.debug("MCP", `Tool '${name}' result: ${stringifySmall(masked)}`);
      return masked;
    } catch (err) {
      logger.error("MCP", `Tool '${name}' errored: ${String(err)}`);
      throw err;
    }
  }

  buildAiTools(tools: MCPTool[]) {
    const map: Record<string, Tool> = {};

    for (const t of tools) {
      const inputSchema = jsonSchema(withIncludeSnapshotDescription(t.inputSchema));
      map[t.name] = tool({
        description: resolvePlaywrightToolDescription(t.name, t.description ?? t.name),
        inputSchema,
        execute: (options: unknown) => {
          const { includeSnapshot, ...rest } = (options as Record<string, unknown>) ?? {};
          return this.callTool(t.name, rest as Record<string, unknown>, {
            includeSnapshot: Boolean(includeSnapshot),
          });
        },
      });
    }

    map["browser_snapshot_and_save"] = this.snapshotAndSaveTool;

    return map;
  }

  private snapshotAndSaveTool = tool({
    description: "Take a snapshot of the current page and save it into a variable.",
    inputSchema: z.object({
      variable: z.string(),
    }),
    execute: async (options) => {
      logger.debug("Tool", `snapshotAndSave args: ${stringifySmall(options)}`);
      const { variable } = options;
      const result = await this.callMcpTool("browser_snapshot", {}, { includeSnapshot: true });

      const textContent = result.content.find((content) => content.type === "text");

      const snapshot = textContent?.text.substring(
        textContent.text.indexOf("```yaml"),
        textContent.text.indexOf("```", textContent.text.indexOf("```yaml") + 6) + 3,
      );

      const snapshotWithoutYaml = snapshot?.replace(/```yaml([\s\S]*?)```/g, "$1");

      if (!snapshotWithoutYaml) {
        logger.warn("Tool", "Couldn't create snapshot.");
        const output = { success: false, reason: "Couldn't create snapshot." } as const;
        logger.debug("Tool", `snapshotAndSave result: ${stringifySmall(output)}`);
        return output;
      }

      setVariable(variable, snapshotWithoutYaml);
      logger.info("Tool", `Snapshot saved to variable '${variable}'`);
      const output = { success: true } as const;
      logger.debug("Tool", `snapshotAndSave result: ${stringifySmall(output)}`);
      return output;
    },
  });

  async getAiTools() {
    const tools = await this.listFilteredTools();
    return this.buildAiTools(tools);
  }

  private removeSnapshots(result: PlaywrightToolOutput): PlaywrightToolOutput {
    try {
      const filteredContent: (TextContent | ImageContent)[] = [];
      for (const c of result.content) {
        if (c.type === "text") {
          const withoutPageSnapshotHeader = c.text.replace(/^- Page Snapshot:/, "");
          const withoutYaml = withoutPageSnapshotHeader.replace(/```yaml[\s\S]*?```/g, "").trim();
          if (withoutYaml) {
            filteredContent.push({ type: "text", text: withoutYaml });
          }
        } else if (c.type === "image") {
          filteredContent.push(c);
        }
      }
      return { ...result, content: filteredContent };
    } catch {
      return result;
    }
  }
}

const playwrightMCP = new PlaywrightMCP();

export const connectPlaywrightMCP = async ({ headless }: { headless?: boolean } = {}) => {
  await playwrightMCP.connect({ headless });
  return playwrightMCP;
};

export const disconnectPlaywrightMCP = async () => {
  await playwrightMCP.disconnect();
};
