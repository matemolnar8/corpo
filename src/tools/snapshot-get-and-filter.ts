import { tool } from "ai";
import z from "zod";
import { parse } from "@std/yaml";
import { getVariable, setVariable } from "./variable.ts";
import { logger } from "../log.ts";

type TextMatch =
  | { equals: string }
  | { contains: string }
  | { regex: string; flags?: string };

type AccessibilityNode = {
  role: string;
  text?: string;
  attributes: Record<string, string>;
  children: AccessibilityNode[];
  rawDescriptor: string;
};

function parseDescriptor(descriptor: string): {
  role: string;
  text?: string;
  attributes: Record<string, string>;
} {
  const attributes: Record<string, string> = {};

  const roleMatch = descriptor.match(/^(\w+)/);
  const role = roleMatch ? roleMatch[1] : "generic";

  const textMatch = descriptor.match(/\s+"([^"]+)"/);
  const text = textMatch ? textMatch[1] : "";

  const attrRegex = /\[([^\]]+)\]/g;
  let m: RegExpExecArray | null;
  // eslint-disable-next-line no-cond-assign
  while ((m = attrRegex.exec(descriptor)) !== null) {
    const part = m[1];
    const eqIndex = part.indexOf("=");
    if (eqIndex === -1) {
      // flag attribute like [checked]
      attributes[part.trim()] = "true";
    } else {
      const key = part.slice(0, eqIndex).trim();
      const rawVal = part.slice(eqIndex + 1).trim();
      const val = rawVal;
      attributes[key] = val.toString();
    }
  }

  return { role, text, attributes };
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function isDescriptorObject(obj: Record<string, unknown>): boolean {
  const keys = Object.keys(obj);
  if (keys.length !== 1) return false;
  const onlyKey = keys[0];
  return !onlyKey.startsWith("/") && Array.isArray(obj[onlyKey]);
}

function buildNodes(input: unknown): AccessibilityNode[] {
  if (typeof input === "string") {
    const parsed = parseDescriptor(input);
    return [
      {
        role: parsed.role,
        text: parsed.text,
        attributes: parsed.attributes,
        children: [],
        rawDescriptor: input,
      },
    ];
  }

  if (Array.isArray(input)) {
    const nodes: AccessibilityNode[] = [];
    for (const item of input) {
      nodes.push(...buildNodes(item));
    }
    return nodes;
  }

  if (isPlainObject(input)) {
    if (isDescriptorObject(input)) {
      const [descriptor, value] = Object.entries(input)[0] as [string, unknown];
      const parsed = parseDescriptor(descriptor);
      const children = Array.isArray(value) ? buildNodes(value) : [];
      return [
        {
          role: parsed.role,
          text: parsed.text,
          attributes: parsed.attributes,
          children,
          rawDescriptor: descriptor,
        },
      ];
    }

    // Property object (e.g., {"/url": "#"}) or unknown; ignore for now
    return [];
  }

  return [];
}

function textMatches(candidate: string | undefined, matcher?: TextMatch): boolean {
  if (!matcher) return true;
  if (candidate == null) return false;
  if ("equals" in matcher) return candidate.toLowerCase().trim() === matcher.equals.toLowerCase().trim();
  if ("contains" in matcher) return candidate.toLowerCase().trim().includes(matcher.contains.toLowerCase().trim());
  if ("regex" in matcher) {
    const re = new RegExp(matcher.regex, matcher.flags);
    return re.test(candidate.toLowerCase().trim());
  }
  return false;
}

function attributesMatch(
  candidate: Record<string, string>,
  required?: Record<string, string>,
): boolean {
  if (!required) return true;
  for (const [k, v] of Object.entries(required)) {
    if (!(k in candidate)) return false;
    if (candidate[k].toLowerCase().trim() !== v.toLowerCase().trim()) return false;
  }
  return true;
}

function findMatches(
  nodes: AccessibilityNode[],
  filter: {
    role?: string | string[];
    text?: TextMatch;
    attributes?: Record<string, string>;
    includeSubtree: boolean;
    mode: "first" | "all";
    maxResults?: number;
  },
): AccessibilityNode[] {
  const roles = Array.isArray(filter.role) ? new Set(filter.role) : filter.role ? new Set([filter.role]) : undefined;
  const results: AccessibilityNode[] = [];

  const visit = (node: AccessibilityNode) => {
    if (
      (!roles || roles.has(node.role)) &&
      textMatches(node.text, filter.text) &&
      attributesMatch(node.attributes, filter.attributes)
    ) {
      const copy: AccessibilityNode = {
        role: node.role,
        text: node.text,
        attributes: { ...node.attributes },
        children: filter.includeSubtree ? node.children.map(cloneNode) : [],
        rawDescriptor: node.rawDescriptor,
      };
      results.push(copy);
      if (filter.mode === "first") return true;
      if (filter.maxResults && results.length >= filter.maxResults) return true;
    }

    for (const child of node.children) {
      const stop = visit(child);
      if (stop) return true;
    }
    return false;
  };

  for (const n of nodes) {
    const stop = visit(n);
    if (stop) break;
  }

  return results;
}

function cloneNode(node: AccessibilityNode): AccessibilityNode {
  return {
    role: node.role,
    text: node.text,
    attributes: { ...node.attributes },
    children: node.children.map(cloneNode),
    rawDescriptor: node.rawDescriptor,
  };
}

export const snapshotGetAndFilterInputSchema = z.object({
  variable: z.string().describe("Name of the variable that holds the YAML accessibility snapshot"),
  filter: z
    .object({
      role: z.union([z.string(), z.array(z.string())]).optional()
        .describe("Filter by accessibility role (e.g., 'table', 'link', 'row')"),
      text: z
        .union([
          z.object({ equals: z.string() }),
          z.object({ contains: z.string() }),
          z.object({ regex: z.string(), flags: z.string().optional() }),
        ])
        .optional()
        .describe(
          "Filter by accessible name/text using equals, contains, or regex.",
        ),
      attributes: z
        .record(z.string(), z.string())
        .optional()
        .describe("Match descriptor attributes like level=4, cursor=pointer"),
    })
    .default({})
    .describe("Filter by accessibility role/text/attributes. Make sure to use the most specific filter possible."),
  includeSubtree: z.boolean().default(false).describe(
    "Include the full subtree of matched nodes. Expensive, should be used when absolutely necessary.",
  ),
  mode: z.enum(["first", "all"]).default("all").describe("Return only the first match or all matches"),
  maxResults: z.number().int().positive().optional().describe("Limit the number of returned matches"),
  storeInVariable: z.string().optional().describe("If provided, store the YAML result in this variable name"),
});

export const snapshotGetAndFilterOutputSchema = z.object({
  success: z.boolean(),
  count: z.number().int().nonnegative(),
  json: z.string().optional(),
  reason: z.string().optional(),
});

export const snapshotGetAndFilterTool = tool({
  description:
    "Parse an accessibility YAML snapshot from a variable and return a filtered subtree by role/text/attributes. Only works when snapshots are saved with the browser_snapshot_and_save tool.",
  inputSchema: snapshotGetAndFilterInputSchema,
  outputSchema: snapshotGetAndFilterOutputSchema,
  execute: ({ variable, filter, includeSubtree, mode, maxResults, storeInVariable }) => {
    const MAX_JSON_CHARS = 30_000;
    // Start log
    let __argsStr = "";
    try {
      __argsStr = JSON.stringify({ variable, filter, includeSubtree, mode, maxResults, storeInVariable });
    } catch {
      __argsStr = String({ variable, filter, includeSubtree, mode, maxResults, storeInVariable });
    }
    logger.debug("Tool", `snapshot_get_and_filter args: ${__argsStr}`);
    const raw = getVariable(variable);
    if (!raw) {
      logger.debug(
        "Tool",
        `snapshot_get_and_filter result: ${
          JSON.stringify({ success: false, count: 0, reason: `Variable '${variable}' not found` })
        }`,
      );
      return { success: false, count: 0, reason: `Variable '${variable}' not found` };
    }

    let parsed: unknown;
    try {
      parsed = parse(raw);
    } catch (err) {
      logger.debug(
        "Tool",
        `snapshot_get_and_filter result: ${
          JSON.stringify({ success: false, count: 0, reason: `Failed to parse YAML: ${(err as Error).message}` })
        }`,
      );
      return { success: false, count: 0, reason: `Failed to parse YAML: ${(err as Error).message}` };
    }

    const rootNodes = buildNodes(parsed);
    const matches = findMatches(rootNodes, {
      role: filter.role,
      text: filter.text,
      attributes: filter.attributes,
      includeSubtree,
      mode,
      maxResults,
    });

    const json = JSON.stringify(matches, null, 2);

    // Enforce size limit to encourage narrower filters
    if (json.length > MAX_JSON_CHARS) {
      const reason = `Filtered result too large (${json.length} > ${MAX_JSON_CHARS} chars). ` +
        "Narrow your filter: restrict 'role'/'text'/'attributes', avoid 'includeSubtree' unless required, or set 'maxResults'.";
      logger.debug(
        "Tool",
        `snapshot_get_and_filter result: ${JSON.stringify({ success: false, count: matches.length, reason })}`,
      );
      return { success: false, count: matches.length, reason };
    }

    if (storeInVariable) {
      setVariable(storeInVariable, json);
    }
    logger.debug(
      "Tool",
      `snapshot_get_and_filter result: ${JSON.stringify({ success: true, count: matches.length, json })}`,
    );
    return { success: true, count: matches.length, json };
  },
});
