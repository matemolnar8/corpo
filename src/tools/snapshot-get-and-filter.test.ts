import { expect } from "@std/expect";
import { snapshotGetAndFilterTool } from "./snapshot-get-and-filter.ts";
import { connectPlaywrightMCP } from "./mcp/playwright-mcp.ts";
import { getVariable, resetVariables, setVariable } from "./variable.ts";
import { assertSnapshot } from "@std/testing/snapshot";
import { setLogLevel } from "../log.ts";

setLogLevel("debug");

Deno.test("snapshot get and filter extracts heading and link from applitools snapshot", async (context) => {
  resetVariables();
  const mcp = await connectPlaywrightMCP({ headless: true });

  await mcp.callTool("browser_navigate", { url: "https://demo.applitools.com/" });
  await mcp.callTool("browser_snapshot_and_save", { variable: "applitools_snapshot" });

  expect(getVariable("applitools_snapshot")).toBeDefined();

  const headingResult = await snapshotGetAndFilterTool.execute!(
    {
      variable: "applitools_snapshot",
      filter: { role: "heading", text: { contains: "Login" } },
      includeSubtree: false,
      mode: "first",
    },
    { messages: [], toolCallId: crypto.randomUUID() },
  );
  const heading = headingResult as { success: boolean; count: number; yaml?: string };
  assertSnapshot(context, heading);

  const linkResult = await snapshotGetAndFilterTool.execute!(
    {
      variable: "applitools_snapshot",
      filter: { role: "link", text: { contains: "Sign in" } },
      includeSubtree: true,
      mode: "first",
    },
    { messages: [], toolCallId: crypto.randomUUID() },
  );
  const link = linkResult as { success: boolean; count: number; yaml?: string };
  assertSnapshot(context, link);

  await mcp.disconnect();
  await new Promise((resolve) => setTimeout(resolve, 1000));
});

Deno.test("snapshot get and filter extracts Cloudbooking link from okta snapshot", async (context) => {
  resetVariables();
  setVariable("okta_snapshot", await Deno.readTextFile("./resources/okta-snapshot.yml"));

  const cloudbookingResult = await snapshotGetAndFilterTool.execute!(
    {
      variable: "okta_snapshot",
      filter: { role: "link", text: { contains: "Cloudbooking" } },
      includeSubtree: true,
      mode: "first",
    },
    { messages: [], toolCallId: crypto.randomUUID() },
  );

  assertSnapshot(context, cloudbookingResult);
});

Deno.test("snapshot get and filter extract bookings from specified date from cloudbooking snapshot", async (context) => {
  resetVariables();
  setVariable("cloudbooking_snapshot", await Deno.readTextFile("./resources/cloudbooking-snapshot.yml"));

  const bookingsResult = await snapshotGetAndFilterTool.execute!(
    {
      variable: "cloudbooking_snapshot",
      filter: { role: "row", text: { contains: "Jul 2025" } },
      includeSubtree: false,
      mode: "all",
    },
    { messages: [], toolCallId: crypto.randomUUID() },
  );

  assertSnapshot(context, bookingsResult);
});

Deno.test("snapshot get and filter enforces size limit and returns failure when too large", async () => {
  resetVariables();
  // Construct a synthetic large snapshot result by creating many similar nodes under a role
  const nodes: unknown[] = [];
  for (let i = 0; i < 2000; i++) {
    nodes.push('row "Item' + i + '"');
  }
  const yaml = `table:\n  - ${nodes.join("\n  - ")}`;
  setVariable("large_snapshot", yaml);

  const result = await snapshotGetAndFilterTool.execute!(
    {
      variable: "large_snapshot",
      filter: { role: "row" },
      includeSubtree: false,
      mode: "all",
    },
    { messages: [], toolCallId: crypto.randomUUID() },
  );

  const { success, reason } = result as { success: boolean; reason?: string };
  expect(success).toBe(false);
  expect(reason).toContain("Filtered result too large");
});
