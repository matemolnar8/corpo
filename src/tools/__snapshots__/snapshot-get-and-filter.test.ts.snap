export const snapshot = {};

snapshot[`snapshot get and filter extracts heading and link from applitools snapshot 1`] = `
{
  count: 1,
  json: '[
  {
    "role": "heading",
    "text": "Login Form",
    "attributes": {
      "level": "4",
      "ref": "e6"
    },
    "children": [],
    "rawDescriptor": "heading \\\\"Login Form\\\\" [level=4] [ref=e6]"
  }
]',
  success: true,
}
`;

snapshot[`snapshot get and filter extracts heading and link from applitools snapshot 2`] = `
{
  count: 1,
  json: '[
  {
    "role": "link",
    "text": "Sign in",
    "attributes": {
      "ref": "e17",
      "cursor": "pointer"
    },
    "children": [],
    "rawDescriptor": "link \\\\"Sign in\\\\" [ref=e17] [cursor=pointer]"
  }
]',
  success: true,
}
`;

snapshot[`snapshot get and filter extracts Cloudbooking link from okta snapshot 1`] = `
{
  count: 1,
  json: '[
  {
    "role": "link",
    "text": "launch app Cloudbooking",
    "attributes": {},
    "children": [
      {
        "role": "article",
        "text": "",
        "attributes": {},
        "children": [
          {
            "role": "img",
            "text": "Cloudbooking logo",
            "attributes": {},
            "children": [],
            "rawDescriptor": "img \\\\"Cloudbooking logo\\\\""
          }
        ],
        "rawDescriptor": "article"
      }
    ],
    "rawDescriptor": "link \\\\"launch app Cloudbooking\\\\""
  }
]',
  success: true,
}
`;

snapshot[`snapshot get and filter extract bookings from specified date from cloudbooking snapshot 1`] = `
{
  count: 14,
  json: '[
  {
    "role": "row",
    "text": "00180877 Desk Complete 29 Jun 2025 22:01 Mate.Molnar@genesys.com Mate Molnar 03 Jul 2025 06:00 03 Jul 2025 21:00 Budapest - Eiffel Square 5th floor - Desk 5-0106",
    "attributes": {},
    "children": [],
    "rawDescriptor": "row \\\\"00180877 Desk Complete 29 Jun 2025 22:01 Mate.Molnar@genesys.com Mate Molnar 03 Jul 2025 06:00 03 Jul 2025 21:00 Budapest - Eiffel Square 5th floor - Desk 5-0106\\\\""
  },
  {
    "role": "row",
    "text": "00181874 Desk Complete 07 Jul 2025 05:11 Mate.Molnar@genesys.com Mate Molnar 07 Jul 2025 06:00 07 Jul 2025 21:00 Budapest - Eiffel Square 5th floor - Desk 5-0106",
    "attributes": {},
    "children": [],
    "rawDescriptor": "row \\\\"00181874 Desk Complete 07 Jul 2025 05:11 Mate.Molnar@genesys.com Mate Molnar 07 Jul 2025 06:00 07 Jul 2025 21:00 Budapest - Eiffel Square 5th floor - Desk 5-0106\\\\""
  },
  {
    "role": "row",
    "text": "00181875 Desk Complete 07 Jul 2025 05:12 Mate.Molnar@genesys.com Mate Molnar 09 Jul 2025 06:00 09 Jul 2025 21:00 Budapest - Eiffel Square 5th floor - Desk 5-0106",
    "attributes": {},
    "children": [],
    "rawDescriptor": "row \\\\"00181875 Desk Complete 07 Jul 2025 05:12 Mate.Molnar@genesys.com Mate Molnar 09 Jul 2025 06:00 09 Jul 2025 21:00 Budapest - Eiffel Square 5th floor - Desk 5-0106\\\\""
  },
  {
    "role": "row",
    "text": "00181876 Desk Complete 07 Jul 2025 05:12 Mate.Molnar@genesys.com Mate Molnar 10 Jul 2025 06:00 10 Jul 2025 21:00 Budapest - Eiffel Square 5th floor - Desk 5-0106",
    "attributes": {},
    "children": [],
    "rawDescriptor": "row \\\\"00181876 Desk Complete 07 Jul 2025 05:12 Mate.Molnar@genesys.com Mate Molnar 10 Jul 2025 06:00 10 Jul 2025 21:00 Budapest - Eiffel Square 5th floor - Desk 5-0106\\\\""
  },
  {
    "role": "row",
    "text": "00182832 Desk Complete 13 Jul 2025 18:52 Mate.Molnar@genesys.com Mate Molnar 14 Jul 2025 06:00 14 Jul 2025 21:00 Budapest - Eiffel Square 5th floor - Desk 5-0106",
    "attributes": {},
    "children": [],
    "rawDescriptor": "row \\\\"00182832 Desk Complete 13 Jul 2025 18:52 Mate.Molnar@genesys.com Mate Molnar 14 Jul 2025 06:00 14 Jul 2025 21:00 Budapest - Eiffel Square 5th floor - Desk 5-0106\\\\""
  },
  {
    "role": "row",
    "text": "00182833 Desk Complete 13 Jul 2025 18:52 Mate.Molnar@genesys.com Mate Molnar 16 Jul 2025 06:00 16 Jul 2025 21:00 Budapest - Eiffel Square 5th floor - Desk 5-0106",
    "attributes": {},
    "children": [],
    "rawDescriptor": "row \\\\"00182833 Desk Complete 13 Jul 2025 18:52 Mate.Molnar@genesys.com Mate Molnar 16 Jul 2025 06:00 16 Jul 2025 21:00 Budapest - Eiffel Square 5th floor - Desk 5-0106\\\\""
  },
  {
    "role": "row",
    "text": "00182834 Desk Complete 13 Jul 2025 18:53 Mate.Molnar@genesys.com Mate Molnar 17 Jul 2025 06:00 17 Jul 2025 21:00 Budapest - Eiffel Square 5th floor - Desk 5-0106",
    "attributes": {},
    "children": [],
    "rawDescriptor": "row \\\\"00182834 Desk Complete 13 Jul 2025 18:53 Mate.Molnar@genesys.com Mate Molnar 17 Jul 2025 06:00 17 Jul 2025 21:00 Budapest - Eiffel Square 5th floor - Desk 5-0106\\\\""
  },
  {
    "role": "row",
    "text": "00184101 Desk Complete 20 Jul 2025 20:57 Mate.Molnar@genesys.com Mate Molnar 21 Jul 2025 06:00 21 Jul 2025 21:00 Budapest - Eiffel Square 5th floor - Desk 5-0106",
    "attributes": {},
    "children": [],
    "rawDescriptor": "row \\\\"00184101 Desk Complete 20 Jul 2025 20:57 Mate.Molnar@genesys.com Mate Molnar 21 Jul 2025 06:00 21 Jul 2025 21:00 Budapest - Eiffel Square 5th floor - Desk 5-0106\\\\""
  },
  {
    "role": "row",
    "text": "00184102 Desk Complete 20 Jul 2025 20:57 Mate.Molnar@genesys.com Mate Molnar 23 Jul 2025 06:00 23 Jul 2025 21:00 Budapest - Eiffel Square 5th floor - Desk 5-0106",
    "attributes": {},
    "children": [],
    "rawDescriptor": "row \\\\"00184102 Desk Complete 20 Jul 2025 20:57 Mate.Molnar@genesys.com Mate Molnar 23 Jul 2025 06:00 23 Jul 2025 21:00 Budapest - Eiffel Square 5th floor - Desk 5-0106\\\\""
  },
  {
    "role": "row",
    "text": "00184103 Desk Complete 20 Jul 2025 20:57 Mate.Molnar@genesys.com Mate Molnar 24 Jul 2025 06:00 24 Jul 2025 21:00 Budapest - Eiffel Square 5th floor - Desk 5-0106",
    "attributes": {},
    "children": [],
    "rawDescriptor": "row \\\\"00184103 Desk Complete 20 Jul 2025 20:57 Mate.Molnar@genesys.com Mate Molnar 24 Jul 2025 06:00 24 Jul 2025 21:00 Budapest - Eiffel Square 5th floor - Desk 5-0106\\\\""
  },
  {
    "role": "row",
    "text": "00185261 Desk Complete 27 Jul 2025 13:15 Mate.Molnar@genesys.com Mate Molnar 28 Jul 2025 06:00 28 Jul 2025 21:00 Budapest - Eiffel Square 5th floor - Desk 5-0106",
    "attributes": {},
    "children": [],
    "rawDescriptor": "row \\\\"00185261 Desk Complete 27 Jul 2025 13:15 Mate.Molnar@genesys.com Mate Molnar 28 Jul 2025 06:00 28 Jul 2025 21:00 Budapest - Eiffel Square 5th floor - Desk 5-0106\\\\""
  },
  {
    "role": "row",
    "text": "00185668 Desk Complete 29 Jul 2025 08:51 Mate.Molnar@genesys.com Mate Molnar 29 Jul 2025 08:45 29 Jul 2025 21:00 Budapest - Eiffel Square 5th floor - Desk 5-0106",
    "attributes": {},
    "children": [],
    "rawDescriptor": "row \\\\"00185668 Desk Complete 29 Jul 2025 08:51 Mate.Molnar@genesys.com Mate Molnar 29 Jul 2025 08:45 29 Jul 2025 21:00 Budapest - Eiffel Square 5th floor - Desk 5-0106\\\\""
  },
  {
    "role": "row",
    "text": "00185263 Desk Complete 27 Jul 2025 13:16 Mate.Molnar@genesys.com Mate Molnar 30 Jul 2025 06:00 30 Jul 2025 21:00 Budapest - Eiffel Square 5th floor - Desk 5-0106",
    "attributes": {},
    "children": [],
    "rawDescriptor": "row \\\\"00185263 Desk Complete 27 Jul 2025 13:16 Mate.Molnar@genesys.com Mate Molnar 30 Jul 2025 06:00 30 Jul 2025 21:00 Budapest - Eiffel Square 5th floor - Desk 5-0106\\\\""
  },
  {
    "role": "row",
    "text": "00185262 Desk Complete 27 Jul 2025 13:16 Mate.Molnar@genesys.com Mate Molnar 31 Jul 2025 06:00 31 Jul 2025 21:00 Budapest - Eiffel Square 5th floor - Desk 5-0106",
    "attributes": {},
    "children": [],
    "rawDescriptor": "row \\\\"00185262 Desk Complete 27 Jul 2025 13:16 Mate.Molnar@genesys.com Mate Molnar 31 Jul 2025 06:00 31 Jul 2025 21:00 Budapest - Eiffel Square 5th floor - Desk 5-0106\\\\""
  }
]',
  success: true,
}
`;
